### PR TITLE
Enhance DataCollector to validate model_reporters functions

### PR DESCRIPTION
**Pull request by peter-kinger**
### Summary
This PR enhances the DataCollector class by adding a comprehensive validation system for model reporters. The main goal is to provide clearer feedback when reporters are misconfigured, while maintaining backward compatibility.

### Motive
The original DataCollector implementation had several limitations:
1. No validation for different types of model reporters
2. Silent failures or unclear error messages when reporters were misconfigured
3. Inconsistent handling of method calls and function parameters
4. No early warning system for potential issues

These issues made debugging difficult and led to confusion when setting up model reporters.

### Implementation
1. New Validation Method:
```python
def _validate_model_reporter(self, name, reporter, model):
    """Validates four types of model reporters:
    1. Lambda functions
    2. Method references (as strings)
    3. Attribute strings
    4. Function lists with parameters
    """
```
2. Modified Collection Logic:
```python
def collect(self, model):
    if self.model_reporters:
        for var, reporter in self.model_reporters.items():
            # Add validation before collection
            self._validate_model_reporter(var, reporter, model)
            # Existing collection logic continues...
```
3. Warning System:
- Replaced hard errors with warning messages
- Added specific warnings for each reporter type
- Included examples in warning messages

### Usage Examples
1. Lambda Functions:
```python
# Valid usage
model_reporters = {
    "Agent Count": lambda m: len(m.agents)
}

# Invalid usage (will show warning)
model_reporters = {
    "Bad Lambda": lambda m: m.nonexistent_attr
}
```
2. Method References:
```python
# Valid usage
model_reporters = {
    "Grid Size": "get_grid_size"  # As string
}

# Invalid usage (will show warning)
model_reporters = {
    "Grid Size": self.get_grid_size  # Direct reference
}
```
3. Attribute Strings:
```python
# Valid usage
model_reporters = {
    "Total Wealth": "total_wealth"
}

# Invalid usage (will show warning)
model_reporters = {
    "Status": "nonexistent_attribute"
}
```
4. Function Lists:
```python
# Valid usage
model_reporters = {
    "Custom": [calculate_metric, [model, param]]
}

# Invalid usage (will show warning)
model_reporters = {
    "Bad Function": ["not_callable", [1, 2]]
}
```

### Additional Notes

1. Test Coverage:
- Added comprehensive test suite in `test_model_reporters.py`
- Tests cover both valid and invalid configurations
- Includes edge cases and error conditions
2. Backward Compatibility:
- All existing code continues to work
- Warnings can be suppressed if desired
- No breaking changes to the API
3. Documentation:
- Updated docstrings with clear examples
- Added warning messages with helpful suggestions
- Included migration guide for existing code
4. Files Modified:
- `datacollection.py`: the only one py code changed in mesa offical repository.
5. Dependencies:
- No new dependencies added
- Uses standard Python `warnings` module

[datacollection_new.zip](https://github.com/user-attachments/files/18372344/datacollection_new.zip)

### CODE REVIEW

**Comment by EwoutH** [comment]
Thanks for the PR, sounds interesting.

What's the performance overhead?

File: mesa/datacollection.py [hunk 1]
```diff
@@ -134,13 +134,88 @@ def __init__(
...
+            try:
+                # Try to call the lambda with a model instance
+                reporter(model)
+            except Exception as e:
+                warnings.warn(
+                    f"Warning: Lambda reporter '{name}' failed: {e!s}\n"
+                    f"Example of valid lambda: lambda m: len(m.agents)",
+                    UserWarning,
+                    stacklevel=2,
+                )
```
Comment (quaquel) [inline mesa/datacollection.py:156]: why issue a warning instead of an exception?

File: mesa/datacollection.py [hunk 2]
```diff
@@ -263,6 +338,9 @@ def collect(self, model):
         """Collect all the data for the given model object."""
         if self.model_reporters:
             for var, reporter in self.model_reporters.items():
+                # Add validation
+                self._validate_model_reporter(var, reporter, model)
+
```
Comment (quaquel) [inline mesa/datacollection.py:343]: Does this imply that the validation is done every single time you try to collect the data? That seems inefficient and overkill.

**Comment by peter-kinger** [comment]
> Thanks for the PR, sounds interesting.
> 
> What's the performance overhead?

Thanks for your attention to the PR. I apologize for the delayed response as I was unfamiliar with Mesa's testing logistics.

Regarding the overhead:

1. Are you referring to the execution time measured by `global_benchmark.py`?

2. As "quaquel" mentioned, the current version of `_validate_model_reporter` performs validation during data collection. Our performance benchmarks show some noticeable overhead.

Future considerations:

1. I've been working on a new version of the validation logic using `_validate_reporter_format` within the `DataCollector.__init__` function. However, this approach presents challenges, particularly with Type 4 reporters (function with parameters), as we can't fully validate them without the `model` parameter.

2. I would greatly appreciate your suggestions on how to proceed, as this has been a challenging issue to resolve.

File: mesa/datacollection.py (hunk 1 above)
Comment (peter-kinger) [reply mesa/datacollection.py:156]: > why issue a warning instead of an exception?


Thanks for the detailed suggestions for my revised codes! During this time, I try to understand them.

About the two questions:

1. Yes, the mechanism of the expectation can avoid serious problems when using the function of the datacollector. As it's my first time adding the validate function, I chose the soft way to validate: just a warning not to influence the mesa code of the original version. 

I had tried to write a new version of codes, which are all replaced with the expectation instead of the warning.
```python
  # Type 1: Lambda function
        if isinstance(reporter, types.LambdaType):
            try:
                reporter(model)
            except Exception as e:
                raise RuntimeError(
                    f"Lambda reporter '{name}' failed validation: {str(e)}\n"
                    f"Example: lambda m: len(m.agents)"
                ) from e

```


2. You're right that validating every data collection could impact performance. I propose two potential solutions:

- (a) make the "`_validate_model_reporter`" function into the optional function, like:

```python
def __init__(

        self,

        model_reporters=None,

        agent_reporters=None,

        agenttype_reporters=None,

        tables=None,

        validate_reporters=False,  # Add new parameter

    ):
```


- (b) Move validation to initialization time:
Put the "`_validate_model_reporter`" function in the `Class DataCollector __init__()` and the "`_validate_model_reporter`" function can be only run once, boosting the running time and lowing the need of performance.

```python
Class DataCollector:
         def __init__():
              xxxxxxx...
         
        if model_reporters is not None:
            for name, reporter in model_reporters.items():
                self._validate_reporter_format(name, reporter) # my new way
                self._new_model_reporter(name, reporter)




```

I wonder what your opinions are. Thanks for your attention to this PR.

**Comment by quaquel** [comment]
I suggest running the validation logic only on the first-ever call to `DataCollector.collect`. This solves both the overhead problem and your issue with type 4 reporters.

**Comment by peter-kinger** [comment]
> I suggest running the validation logic only on the first-ever call to `DataCollector.collect`. This solves both the overhead problem and your issue with type 4 reporters.

Thanks for the suggestions for the position where the "validate function" puts. I have completed the v2, version two, codes of it:

1. Improve: Due the essential input parameter "model", I put the `**_validate_model_reporter**` within the `collect` function. But it will be validate once.  (put the change of signal firstly avoid losing efficacy in `_validate_model_reporter`)
2.  overhead: add the bool signal of "`self._validated`" in `_init_` to make sure there will be once to be validated in the progress of datacollecting.


In addition, I want to clarify the problems in checks:

1. [run-benchmarks](https://github.com/projectmesa/mesa/actions/runs/12705394103/job/35416342272?pr=2605#logs): The benchmark failure (ModuleNotFoundError: No module named 'scipy') appears to be a pre-existing issue with the test environment's dependencies, unrelated to my changes in `datacollector.py`.
2. examples: new version codes have passed the unit test of pytest. The previous fail of unnecessary positional argument had solved.
3. Codecov: I have added corresponding unit tests to cover the new components introduced in `_validate_model_reporter`, ensuring adequate test coverage.

I don't know if my modification is qualified, thank you for your review.

**Comment by quaquel** [comment]
Thanks for the update. I'll try to review the code later today. I did update your branch by merging the main back into it, so everything is up to date again.

1. Yes, scipy was an issue that has, in the meantime, been resolved. You can also see that all tests have now been passed.
2. The one outstanding check that is failing is the codecoverage one. Ideally, we want PRs to be as close as reasonably possible to 100% coverage. The reason your coverage is lower than desired is that you haven't checked the raise exception clauses in your test code. You can do this with `pytest.raises`. It would be great if you can explicitly add tests for this.

**Comment by peter-kinger** [comment]
Thank you very much for taking the time to review all the code revisions in detail. I sincerely apologize for the delayed response due to the Chinese New Year holiday.

I greatly appreciate your suggestion regarding the use of `pytest.raises`. I will add the corresponding test cases in the coming days to ensure 100% test coverage.

**Comment by quaquel** [comment]
This looks ready to go. @peter-kinger, is there any outstanding issue, or can I do a final review and merge this?

**Comment by peter-kinger** [comment]
Hi  @quaquel, based on my observation, there aren't any outstanding issues. I'm not familiar with the performance benchmarks provided by GitHub Actions. If the benchmarks pass, I think we can merge it

In addtion, issues/2606(https://github.com/projectmesa/mesa/issues/2606) should be closed if the PR finishs. Thanks for the reviewing!

**Review by quaquel — APPROVED** [review]

**Comment by quaquel** [comment]
Thanks!
